src/withStyles.jsx

```diff
@@ -51,28 +51,77 @@ export function withStyles(
   /** Cache for storing the result of stylesFn(theme) for all themes. */
   const stylesFnResultCacheMap = typeof WeakMap === 'undefined' ? new Map() : new WeakMap();
 
+  function getOrCreateStylesFnResultCache(theme) {
+    // Get and store the result in the stylesFnResultsCache for the component
+    // -- not the instance -- so we only apply the theme to the stylesFn
+    // once per theme for this component.
+    const cachedResultForTheme = stylesFnResultCacheMap.get(theme);
+    const stylesFnResult = cachedResultForTheme || stylesFn(theme) || {};
+    stylesFnResultCacheMap.set(theme, stylesFnResult); // cache the result of stylesFn(theme)
+    return stylesFnResult;
+  }
+
+  /** Cache for storing the results of computations, theme -> WithStyles -> direction -> results */
+  const withStylesCache = new Map();
+
+  function getComponentCache(theme, component) {
+    const themeCache = withStylesCache.get(theme);
+    if (!themeCache) {
+      return null;
+    }
+    const componentCache = themeCache.get(component);
+    if (!componentCache) {
+      return null;
+    }
+    return componentCache;
+  }
+
+  function getOrCreateComponentCache(theme, component) {
+    let themeCache = withStylesCache.get(theme);
+    if (!themeCache) {
+      themeCache = new Map();
```

Comment:
perhaps this should be:
```suggestion
      themeCache = new (WeakMap || Map)();
```